src/python/m5/params.py

```diff
@@ -897,71 +908,70 @@ def pybind_predecls(cls, code):
-                        if xor_high_bit != 0:
-                            bit2 = xor_high_bit - i
-                            mask |= 1 << bit2
-                        self.masks[self.intlvBits - i - 1] = mask
-
-        if len(args) == 0:
-            self.start = Addr(kwargs.pop("start"))
-            handle_kwargs(self, kwargs)
-
-        elif len(args) == 1:
-            if kwargs:
-                self.start = Addr(args[0])
-                handle_kwargs(self, kwargs)
-            elif isinstance(args[0], (list, tuple)):
-                self.start = Addr(args[0][0])
-                self.end = Addr(args[0][1])
-            else:
+        # Handle getting the start and end addresses
+        match (start, end, size):
+            # case where start is a list/tuple
+            case ([start, end] | (start, end), None, None):
+                self.start = Addr(start)
+                self.end = Addr(end)
+            # cases where start is an int/str/None
+            case (start, None, size) if size is not None:
+                self.start = Addr(start or 0)
+                self.end = self.start + Addr(size)
+            case (start, end, _) if end is not None:
+                self.start = Addr(start or 0)
+                self.end = Addr(end)
+            # case where only one address is specified (given as start but will be used as end)
+            case (start, None, None) if start is not None:
                 self.start = Addr(0)
-                self.end = Addr(args[0])
-
-        elif len(args) == 2:
-            self.start = Addr(args[0])
-            self.end = Addr(args[1])
-        else:
-            raise TypeError("Too many arguments specified")
+                self.end = Addr(start)
+            # failure case
+            case params:
+                raise TypeError(
+                    f"Either end or size must be specified: {params}"
+                )
 
-        if kwargs:
-            raise TypeError(f"Too many keywords: {list(kwargs.keys())}")
+        # now on to the optional bit
+        if intlvMatch is not None:
+            self.intlvMatch = int(intlvMatch)
+
+        if masks is not None:
+            self.masks = [int(x) for x in list(masks)]
+            self.intlvBits = len(self.masks)
+        elif intlvBits is not None:
+            self.intlvBits = int(intlvBits)
+            self.masks = [0] * self.intlvBits
+            if intlvHighBit is None:
+                raise TypeError("No interleave bits specified")
+            intlv_high_bit = int(intlvHighBit)
+            xor_high_bit = 0
+            if xorHighBit is not None:
+                xor_high_bit = int(xorHighBit)
+            for i in range(0, self.intlvBits):
+                bit1 = intlv_high_bit - i
+                mask = 1 << bit1
+                if xor_high_bit != 0:
+                    bit2 = xor_high_bit - i
+                    mask |= 1 << bit2
+                self.masks[self.intlvBits - i - 1] = mask
 
```

Comment:
I'm pretty highly confident this is a correct implementation, but I would appreciate another set of eyes